libraries/ruby_install.rb

```diff
@@ -0,0 +1,252 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+require_relative '_helper'
+
+class Chef
+  class Resource::RubyInstall < Resource::LWRPBase
+    self.resource_name = :ruby_install
+
+    actions :install
+    default_action :install
+
+    attribute :version,     kind_of: String, name_attribute: true
+    attribute :environment, kind_of: Hash, default: {}
+    attribute :patches,     kind_of: Array, default: []
+    attribute :prefix,      kind_of: String, default: lazy { |r| ChefConfig.windows? ? ::File.join(ENV['SYSTEMDRIVE'], 'rubies', r.version) : "/opt/rubies/ruby-#{r.version}" }
+
+    def patch(patch)
+      @patches << patch
+    end
+  end
+
+  class Provider::RubyInstallUnix < Provider::LWRPBase
+    provides :ruby_install
+
+    def whyrun_supported?
+      true
+    end
+
+    action(:install) do
+      if installed?
+        Chef::Log.debug("#{new_resource} installed - skipping")
+      else
+        converge_by("install #{new_resource}") do
+          install
+        end
+      end
+    end
+
+    def self.compile_flags
+      [
+        '--disable-install-rdoc',
+        '--disable-install-ri',
+        '--with-out-ext=tcl',
+        '--with-out-ext=tk',
+        '--without-tcl',
+        '--without-tk',
+        '--disable-dtrace',
+      ].join(' ')
+    end
+
+    protected
+
+    def version
+      new_resource.version
+    end
+
+    def install
+      install_dependencies
+
+      # Need to compile the command outside of the execute resource because
+      # Ruby is bad at instance_eval
+      install_command = "ruby-install --no-install-deps --install-dir #{new_resource.prefix}"
+
+      new_resource.patches.each do |p|
+        install_command << " --patch #{p}"
+      end
+
+      install_command << " ruby #{version} -- #{Provider::RubyInstallUnix.compile_flags}"
+
+      execute = Resource::Execute.new("install ruby-#{version}", run_context)
+      execute.command(install_command)
+      execute.environment(new_resource.environment)
+      execute.run_action(:run)
+    end
+
+    # Check if the given Ruby is installed in the given prefix.
+    #
+    # @return [true, false]
+    def installed?
+      ::File.directory?(new_resource.prefix)
+    end
+
+    def install_dependencies
+      # install build-essentials
+      package = Chef::Resource::Package.new('build-essential', run_context)
+      package.package_name('build-essential')
+      package.run_action(:install)
+
+      # install ruby-install
+      return if Chef::Sugar::Shell.installed_at_version?('ruby-install', '0.4.1')
```

Comment:
Let's do this as a `not_if` guard on the resource...we should also fully qualify the path to `ruby-install`:

```ruby
ruby_install.not_if { installed_at_version?('/usr/local/bin/ruby-install', '0.4.1') }
```